### ISSUE TITLE
fyne package -os android needs NDK 16/19c

### ISSUE THREAD

**Issue by bon-ami** [fyne-io/fyne#3066]
### Describe the bug:

#2922 & #2919 related but both closed. How to build android app now?
I have downloaded NDK 16, 19 & 24 in Android Studio. However armv7a-linux-androideabi16-clang still does not exist. armv7a-linux-androideabi19-clang exists.

### To Reproduce:
Steps to reproduce the behaviour:
1. `fyne package -os android`

### Screenshots:

`no compiler for arm was found in the NDK (tried D:\SDK\ndk\24.0.8215888\toolchains\llvm\prebuilt\windows-x86_64\bin\armv7a-linux-androideabi16-clang). Make sure your NDK version is >= r19c. Use sdkmanager --update to update it`

### Device (please complete the following information):
 - **OS:** Windows
 - **Version:** 10
 - **Go version:** 1.18
 - **Fyne version:** 2.1.4

**Comment by andydotxyz** [comment]
What is the output of "fyne --version"?

**Comment by bon-ami** [comment]
@andydotxyz , I've just upgraded fyne from 2.1.4 to **2.2.1** and the failure persists.

**Comment by bon-ami** [comment]
@andydotxyz , I've just upgraded fyne from 2.1.4 to 2.2.1 and the failure persists.

**Comment by andydotxyz** [comment]
> What is the output of "fyne --version"?

Can you please confirm this?

**Comment by bon-ami** [comment]
> 

@andydotxyz , yes. Now it passes with
`$ fyne --version`
`fyne version v2.2.1`

**Comment by andydotxyz** [comment]
Excellent, it sounds like you did not have the bug fix that we recently released, I am glad this is resolved.

**Comment by Sanichka** [comment]
Hi @andydotxyz 
For some reason I'm still expiriensing same issue as bon-ami while using fyne version v2.2.3 during packaging for android. I've tried NDK 25.0.8775105 from adnroid stuidio tools>sdk manager and also ndk r24 from [this repo](https://github.com/android/ndk/wiki/Unsupported-Downloads) of unsupported downloads of ndk.  Could you please advice if there anything that might lead to the same issue?
![image](https://user-images.githubusercontent.com/9090737/183934238-8d6c4c25-d491-44cf-a9f7-4b1bf04ed116.png)

**Comment by andydotxyz** [comment]
Not really resolved yet it seems

**Comment by Sanichka** [comment]
Thank you for reply. I'm looking forward for any news and ready to test and cooperate to solve this issue!

**Comment by andydotxyz** [comment]
Thanks @Sanichka, you could try out the proposed fix at #3219

**Comment by Sanichka** [comment]
Hi @andydotxyz 
I've pulled fix and replaced fyne in go.mod to local folder with fix, but either I've done something wrong or it didn't help considering same error message. Is there anything wrong with my usage of fixed version or anything I've missed?
![fyne1](https://user-images.githubusercontent.com/9090737/186635158-53ceea09-3cd9-454d-a6e7-941f3b294d93.png)

**Comment by Bluebugs** [comment]
You actually need to pull the binary from that PR. I am not to sure if there is a best way, but I think you have to checkout that PR, then build and install manually the fyne binary with go build/install from the cmd/fyne subdirectory.

**Comment by Sanichka** [comment]
Thank you for hint, I've managed to build and tried pasting fyne.exe in my project folder and packaging and replaced old exe in `$GOPATH/bin` and same error there. P.S: My module called qrscannerapp and it's seems it's missing some file during packagin process. Hope this info will help, let me know if you need anything else.
![image](https://user-images.githubusercontent.com/9090737/186729808-4527e055-1c32-4fd3-8688-67bce339dd74.png)

**Comment by Bluebugs** [comment]
Not to sure how to debug this one. On linux you can trace what command is being executed by using `strace -fF -e execve fyne package ...`. I do not know if there is a similar solution on Windows, if you know any that will be helpful.

Otherwise which version of the NDK is installed? And is your application otherwise pure go?

**Comment by Sanichka** [comment]
NDK is r34 from [this ](https://github.com/android/ndk/wiki/Unsupported-Downloads) repo. Yes, for now main.go contains fyne example with hello world:
```
package main

import (
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/widget"
)

func main() {
	a := app.New()
	w := a.NewWindow("Hello World")

	w.SetContent(widget.NewLabel("Hello World!"))
	w.ShowAndRun()
}
```

**Comment by andydotxyz** [comment]
Oh my - `exec: no command found` is it possible there is a corner case the PR did not cover @Bluebugs ?
I wonder if it is looping through possible clangs and not finding one leaving a "" command which would result in that error?...

**Comment by andydotxyz** [comment]
@Sanichka just out of interest, why are you using unsupported downloads repo?
If you install the SDK from "Android Studio" instead does it work correctly?

**Comment by Sanichka** [comment]
I've tried NDK from Android Studio at start, but then switched to unofficial source to test it since some mobile dev related YT videos were using it with go. Should I try version from android studio for testing purposes?

**Comment by Bluebugs** [comment]
> Oh my - `exec: no command found` is it possible there is a corner case the PR did not cover @Bluebugs ? I wonder if it is looping through possible clangs and not finding one leaving a "" command which would result in that error?...

Oh, that could be an explanation. It should error, but maybe it doesn't.

**Comment by Bluebugs** [comment]
> I've tried NDK from Android Studio at start, but then switched to unofficial source to test it since some mobile dev related YT videos were using it with go. Should I try version from android studio for testing purposes?

Oh, what kind of NDK are you using?

**Comment by Sanichka** [comment]
Just tried with NDK 25.0.8775105 from Android Studio. Same error. So both versions r24 from unofficial repo and android studio one leads to this.
![fyne3](https://user-images.githubusercontent.com/9090737/188814853-9de774aa-728c-4104-84ec-6083c5dd09c6.png)

**Comment by andydotxyz** [comment]
Confirmed, it silently returns "" as the command if we don't find a matching compiler... https://github.com/fyne-io/fyne/blob/develop/cmd/fyne/internal/mobile/env.go#L384

@Sanichka can you help by looking in your ANDROID_NDK_HOME and seeing if the following file exists? If not, what is the closest match:

`/toolchains/llvm/prebuilt/darwin-x86_64/bin/x86_64-linux-android21-clang`

**Comment by Bluebugs** [comment]
@Sanichka can you give us the full path of one of those files?

**Comment by Sanichka** [comment]
C:\Users\Alex\AppData\Local\Android\Sdk\ndk\25.0.8775105\toolchains\llvm\prebuilt\windows-x86_64\bin\x86_64-linux-android21-clang

**Comment by MoLuoPro** [comment]
I got the same problem:    
no compiler for arm was found in the NDK (tried D:\Cache\AS\SDK\ndk\25.1.8937393\toolchains\llvm\prebuilt\windows-x86_64\bin\armv7a-linux-androideabi16-clang). Make sure your NDK version is >= r19c. Use `sdkmanager --update` to update it

**Comment by Bluebugs** [comment]
@MoLuoPro I am correct into assuming you tried it with the develop branch of fyne? If so, I wonder if the issue is not with my code and windows as I tested it on Linux. Maybe something I overlooked, I will check if I see anything.

**Comment by andydotxyz** [comment]
@Sanichka @MoLuoPro as we prepare for next release plans I want to confirm if you were able to test with the `fyne` command from `develop` branch and whether the problem persisted?

**Comment by Sanichka** [comment]
I've just tested develop branch and I've got same error:
`C:\Users\Alex\AppData\Local\Temp\fyne-work-4222060171\lib\armeabi-v7a\libqrscannerapp.so: exec: no command`

**Comment by Dragon-Hacking** [comment]
> ### Describe the bug:
> 
> #2922 & #2919 related but both closed. How to build android app now? I have downloaded NDK 16, 19 & 24 in Android Studio. However armv7a-linux-androideabi16-clang still does not exist. armv7a-linux-androideabi19-clang exists.
> ### To Reproduce:
> 
> Steps to reproduce the behaviour:
> 
>     1. `fyne package -os android`
> 
> 
> ### Screenshots:
> 
> `no compiler for arm was found in the NDK (tried D:\SDK\ndk\24.0.8215888\toolchains\llvm\prebuilt\windows-x86_64\bin\armv7a-linux-androideabi16-clang). Make sure your NDK version is >= r19c. Use sdkmanager --update to update it`
> ### Example code:
> ### Device (please complete the following information):
> 
>     * **OS:** Windows
> 
>     * **Version:** 10
> 
>     * **Go version:** 1.18
> 
>     * **Fyne version:** 2.1.4

**Comment by Bluebugs** [comment]
Would you mind trying with the latest develop branch out should include a fix for this. I haven't tried it on windows, so maybe there is still some improvement to be made there.

**Comment by Dragon-Hacking** [comment]
**They must go to the directory where the binary is missing in my case** _/lib/android-ndk/android-ndk-r25b/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi16-clang++_

`$ cd $ANDROID_NDK_HOME/toolchains/llvm/prebuilt/linux-x86_64/bin/`

**I copied the binary** "armv7a-linux-androideabi19-clang++" **into a new file which is the one that does not exist** "armv7a-linux-androideabi16-clang++" **repeated in process with the other missing binaries**

`$ cp armv7a-linux-androideabi19-clang++   armv7a-linux-androideabi16-clang++`
`$ cp armv7a-linux-androideabi19-clang   armv7a-linux-androideabi16-clang`

`$ cp i686-linux-android19-clang++   i686-linux-android16-clang++`
`$ cp i686-linux-android19-clang++   i686-linux-android16-clang++`

**I have solved the problem**
![Captura de pantalla -2022-10-01 22-04-06](https://user-images.githubusercontent.com/90647351/193437337-2e3314bd-c60e-433c-b7cb-5c618728e3ce.png)
![a](https://user-images.githubusercontent.com/90647351/193437375-c66fd28c-13c7-4783-b3af-5592fe47a1da.png)

**Comment by andydotxyz** [comment]
@Bluebugs does this data help?
I'm worried this may be a blocking issue, though it is still marked as unconfirmed...

**Comment by Bluebugs** [comment]
I should try to setup a windows machine to see as the informations here are not really shining a light on the problem.

**Comment by observerss** [comment]
same "exec: no command" on win10, fyne 2.4.2,  ndk 25
I find a way to make a build:

1. download ndk 19c manually from https://github.com/android/ndk/wiki/Unsupported-Downloads and extract to somewhere
2. export ANDROID_NDK_HOME=PATH_TO_NDK_19c
3. git clone https://github.com/fyne-io/fyne; git checkout develop; go build -o fyne-develop.exe cmd/fyne/main.go
4. fyne-develop.exe package -os android/amd64 -appID XXX  -icon Icon.png

note that the value of `-os` must be `android/amd64`, if downloaded ndk is for `x86_64`
develop branch actually works, but with a misleading error message when arch is mismatch

**Comment by andydotxyz** [comment]
Can you confirm why you edited this to remove "develop branch actually works" and your steps to replicate @observerss ?

**Comment by tamaraspringer** [comment]
Hello there,
I have a similar issue and I didn't quite understand if this is solved or not. I'm stuck with the `AppData\Local\Temp\fyne-work-150880070\lib\armeabi-v7a\libhello.so: exec: no command`. 
What does this mean and how can I fix that? 
Thanks in advance

**Comment by andydotxyz** [comment]
@Bluebugs can we get to the bottom of this before v2.3.0? I'm not sure how well fixed this problem is...

**Comment by Bluebugs** [comment]
Indeed, I am planning to investigate this week.

**Comment by andydotxyz** [comment]
Cool thanks

**Comment by darkencrow** [comment]
### Maybe this could be solve it.

I got the same issue with this ` exec: no command` error when packaging android apk .

And I found this is because the program try to extract the params from ndk dir but failed within filename  `llvm-nm`, `os.Stat()` func return error with it. So I try to append the ext .`exe` on it, and then everthing's right.

The relating code is located at _%GOPATH%\src\fyne.io\fyne\v2\cmd\fyne\internal\mobile\build_androidapp.go_ -> **goAndroidBuild()** line 94 : `extractPkgs(toolchain.Path(ndkRoot, "nm")...`
And _%GOPATH%\src\fyne.io\fyne\v2\cmd\fyne\internal\mobile\env.go_ -> `(tc *ndkToolchain) Path()` line 380

(ver 2.2.4)

**Wish these can help someone to fix this prob.**

**Comment by Bluebugs** [comment]
Thanks, I will try to have a PR going today for this.

**Comment by Bluebugs** [comment]
I did it slightly differently, just in case, to be a bit more future proof. Have a look.

**Comment by Bluebugs** [comment]
This should be fixed in develop now. Please confirm and we can hopefully close this issue.

**Comment by darkencrow** [comment]
Got it! thank you very much!

**Comment by andydotxyz** [comment]
Excellent, sounds fixed :)

**Comment by hkparker** [comment]
I just experienced this issue trying to switch from `fyne-cross` to `fyne package`.  Using arch and installing the 
`android-ndk` package, then setting up my environment variables correctly, I was able to get to

``no compiler for arm was found in the NDK (tried /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi16-clang). Make sure your NDK version is >= r19c. Use `sdkmanager --update` to update it``

As seen up thread, I made copies of these files from other existing versions (in my case, 21, and I had to do that for both armv7a and i686).  Doing so I was able to move on.

But no I've gotten to the point where I'm missing another file

`fork/exec /opt/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/arm-linux-androideabi-nm: no such file or directory`

Looks like this is #2498 and I'm not sure where to go from here.

I'm not familiar enough with andorid's build environment but it seems like perhaps there should be a way for me to specify an API level that fyne should use that exists on my system?  Or maybe I'm just missing some package, but if so it's hard to find.
